### DISCUSSION TITLE
Support narrowing using `in` keyword

### DISCUSSION THREAD

**Discussion by kumaraditya303** [microsoft/pyright#2596] [Ideas]
See this code:
```py
# pyright: strict


def func(browser: object):
    if isinstance(browser, str):
        reveal_type(browser)
    if browser in ("chrome", "firefox"):
        reveal_type(browser)

```

Pyright is able to narrow in first case of `isinstance` but not in second case. This is very popular pattern in python so worth to be considered.

Objects which should be supported in initial implementation:
- `set`
- `tuple`
- `list`
- `deque`

**Comment by erictraut** [comment]
FWIW, mypy doesn't currently provide narrowing support for this pattern. There is an [open issue](https://github.com/python/mypy/issues/10977), but it hasn't received many upvotes yet. Here is [another related feature request](https://github.com/python/mypy/issues/9718).

We'll wait to see if others upvote this request. In the meantime, you can implement your own user-defined type guard function using the PEP 647 functionality.
